RetailCoder.VBE/App.cs

```diff
@@ -56,36 +57,104 @@ public class App : IDisposable
             _configService = configService;
             _appMenus = appMenus;
             _stateBar = stateBar;
-            _hook = hook;
+            _indenter = indenter;
+            _hooks = hooks;
             _logger = LogManager.GetCurrentClassLogger();
 
-            _hook.KeyPressed += _hook_KeyPressed;
+            _hooks.MessageReceived += hooks_MessageReceived;
             _configService.SettingsChanged += _configService_SettingsChanged;
             _parser.State.StateChanged += Parser_StateChanged;
             _stateBar.Refresh += _stateBar_Refresh;
         }
 
-        private void _stateBar_Refresh(object sender, EventArgs e)
+        private Keys _firstStepHotKey;
+        private bool _isAwaitingTwoStepKey;
+        private bool _skipKeyUp;
+
+        private async void hooks_MessageReceived(object sender, HookEventArgs e)
+        {
+            if (sender is LowLevelKeyboardHook)
+            {
+                if (_skipKeyUp)
+                {
+                    _skipKeyUp = false;
+                    return;
+                }
+
+                if (_isAwaitingTwoStepKey)
+                {
+                    // todo: use _firstStepHotKey and e.Key to run 2-step hotkey action
+                    if (_firstStepHotKey == Keys.I && e.Key == Keys.M)
```

Comment:
*cough* for some values of "thoroughly broken" :wink:

@Vogel612 that's precisely the plan :smile: ...I happened to bump into some difficulties putting it all together. Hotkeys are correctly captured and update the "status bar", but the Ctrl/Shift modifiers "stick" and the 2nd key becomes impossible to capture, not to mention how annoying it becomes to have Ctrl/Shift "stuck". As for the low-level keyboard hook, it works but tends to destabilize the whole environment and has been identified as a source of unexpected crashes.